### PR TITLE
Refactor set_agg and set_union function

### LINKED ISSUE #21362: Use @TypeParameter to simplify set_agg aggregation function
With the merge of #21316 we can now simplify this aggregation function implementation by using @TypeParameter annotations in the serializer and state factory classes, which means the function can be entirely converted to using the [function annotation framework](http://prestodb.io/docs/0.284/develop/functions.html). This will simplify the implementation and improve maintainability.

### PR DESCRIPTION
**Pull request by tailemanh**
## Description
Resolve #21362 & #21363. This commit refactors `set_agg` and `set_union` function using the [function annotation framework](http://prestodb.io/docs/0.284/develop/functions.html) 

## Motivation and Context
This will simplify the implementation and improve maintainability

## Impact
None.

## Test Plan
Unit tests have been written before

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by manhtaile** [comment]
Hi @ZacBlanco, I created this PR that resolve the `set_agg` function (#21362), it doesn't pass the ci/cd test, but the failed test involves the `set_union` function (#21363), even though I did nothing to that function (except that they share the same `SetAggregationStateFactory` and `SetAggregationStateSerializer`).
I want to ask, can we combine two tickets handling `set_agg` and `set_union` into one? or is there any other solution? Btw, I'm also willing to contribute to both of these. Thanks!

**Comment by ZacBlanco** [comment]
Hi @manhtaile, you're correct that the error does have to do with the function. I took a look at one of the failures:

```

java.lang.AssertionError: Execution of 'actual' query failed: select group_id, set_union(numbers) from (values (1, array[1, 2]), (1, array[2, 3]), (2, array[4, 5]), (2, array[5, 6])) as t(group_id, numbers) group by group_id
	at org.testng.Assert.fail(Assert.java:98)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:178)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:106)
	at com.facebook.presto.tests.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:157)
	at com.facebook.presto.tests.AbstractTestQueries.testSetUnion(AbstractTestQueries.java:5818)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:135)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:673)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:220)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:945)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:193)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.RuntimeException: com.facebook.presto.common.block.ArrayBlockBuilder
	at com.facebook.presto.tests.AbstractTestingPrestoClient.execute(AbstractTestingPrestoClient.java:124)
	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:733)
	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:701)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:175)
	... 18 more
Caused by: java.lang.UnsupportedOperationException: com.facebook.presto.common.block.ArrayBlockBuilder
	at com.facebook.presto.common.block.BlockBuilder.writeInt(BlockBuilder.java:43)
	at com.facebook.presto.common.block.SingleArrayBlockWriter.writeInt(SingleArrayBlockWriter.java:92)
	at com.facebook.presto.common.type.AbstractIntType.appendTo(AbstractIntType.java:84)
	at com.facebook.presto.operator.aggregation.SetOfValues.serialize(SetOfValues.java:82)
	at com.facebook.presto.operator.aggregation.arrayagg.SetAggregationStateSerializer.serialize(SetAggregationStateSerializer.java:48)
	at com.facebook.presto.operator.aggregation.arrayagg.SetAggregationStateSerializer.serialize(SetAggregationStateSerializer.java:25)
	at com.facebook.presto.$gen.Array_inteIntegerSetUnionGroupedAccumulator_20231114_103852_38151.evaluateIntermediate(Unknown Source)
	at com.facebook.presto.operator.aggregation.builder.InMemoryHashAggregationBuilder$Aggregator.evaluate(InMemoryHashAggregationBuilder.java:466)
	at com.facebook.presto.operator.aggregation.builder.InMemoryHashAggregationBuilder.lambda$buildResult$3(InMemoryHashAggregationBuilder.java:346)
	at com.facebook.presto.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:315)
	at com.facebook.presto.operator.HashAggregationOperator.getOutput(HashAggregationOperator.java:463)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:428)
	at com.facebook.presto.operator.Driver.lambda$processFor$9(Driver.java:311)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:732)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:304)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1079)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:165)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:621)
	at com.facebook.presto.$gen.Presto_0_285_SNAPSHOT_ad2cb15__testversion____20231114_103136_28052.run(Unknown Source)
```

You may need to do some digging to compare the the type of BlockBuilder passed to the output function in the previous version before your updates, and compare it to the one with the annotation framework. If they are different, they you probably need to update some of the logic in the serialization function, or update the return type of the function.

If you get stuck on this let me know and I can look a bit deeper

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/arrayagg/SetAggregationFunction.java [hunk 1]
```diff
@@ -13,124 +13,32 @@
...
-    private static List<ParameterMetadata> createInputParameterMetadata(Type value)
+    private SetAggregationFunction()
     {
-        return ImmutableList.of(new ParameterMetadata(STATE), new ParameterMetadata(NULLABLE_BLOCK_INPUT_CHANNEL, value), new ParameterMetadata(BLOCK_INDEX));
     }
 
-    public static void input(Type type, SetAggregationState state, Block block, int position)
+    @InputFunction
+    @TypeParameter("T")
+    public static void input(@TypeParameter("T") Type type, @AggregationState SetAggregationState state, @BlockPosition @SqlType("T") @NullablePosition Block block, @BlockIndex int position)
```
Comment (ZacBlanco) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/arrayagg/SetAggregationFunction.java:41]: nit: For readability, I'd recommend breaking this up by putting each argument on its own line.

**Comment by ZacBlanco** [comment]
> I want to ask, can we combine two tickets handling `set_agg` and `set_union` into one? or is there any other solution? Btw, I'm also willing to contribute to both of these. Thanks!

I think it would be best to keep separate PRs in case any issues come up afterwards. It will help isolate any breaking issues and let us revert only the breaking code change

**Comment by manhtaile** [comment]
> I think it would be best to keep separate PRs in case any issues come up afterwards. It will help isolate any breaking issues and let us revert only the breaking code change

I just reviewed the code of `SetAggregationFunction` and `SetUnionFunction`, perhaps the cause comes from the difference between the parameters passed to the `SetOfValues` constructor in the input function - in `set_agg` we use ` @TypeParameter("T")` and `set_union` are not. I actually think implementing a function annotation for `set_union` would solve the problem here

**Comment by ZacBlanco** [comment]
Can you please fill out form to authorize your contribution as well? You can find the instructions in the [comment from the CLA bot](https://github.com/prestodb/presto/pull/21378#issuecomment-1809714307)

**Comment by tdcmeehan** [comment]
Keeping these in one PR is fine, the most important thing is to keep them as separate commits so that they have distinct entries in the commit history.

@tailemanh is this ready to review?

**Comment by tdcmeehan** [comment]
@tailemanh then please open this PR--it is still marked as draft.

**Review by tdcmeehan — CHANGES_REQUESTED** [review]
Please squash commits and ensure that all changes for `set_agg` go into one commit, and `set_union` go into a separate distinct commit.

Please ensure the commit messages follow our guidelines, namely:

* Capitalize the first letter: `refactor` => `Refactor`
* Squash the `fix checkstyle ` commit

**Review by ZacBlanco — APPROVED** [review]

**Review by tdcmeehan — APPROVED** [review]
Thank you for simplifying this code @tailemanh!

**Pull request merged**